scipy/linalg/_decomp_lu.py

```diff
@@ -174,53 +182,166 @@ def lu(a, permute_l=False, overwrite_a=False, check_finite=True):
         Whether to check that the input matrix contains only finite numbers.
         Disabling may give a performance gain, but may result in problems
         (crashes, non-termination) if the inputs do contain infinities or NaNs.
+    matrix_p : bool, optional
+        If ``False`` the permutation information is returned as row indices.
+        The default is ``True`` for backwards-compatibility reasons.
```

Comment:
From what I can tell based on the docstring example, we're not deprecating one variant for the other here? As in: both the vector of row indices as well as the full permutation matrix will be useful, depending on the usecase?

In that case, I don't think we should talk about backwards-compat at all - we're not planning to deprecate or remove anything (if both are useful). This would also allow us to avoid "leading" with `use False to do X`, which is the sort of inverted logic that's hard to remember. 

Also, I think we need to note the dependence on `permute_l` here, because this option makes no sense if `permute_l=True`. Perhaps such a combination should even error?

Perhaps something like the following?
```suggestion
    matrix_p : bool, optional
        If ``True`` (the default), return permutation information as a matrix;
        if ``False``, return permutation information as row indices.
        This option only takes effect if ``permute_l=False``.
```

Reply:
reworded